event/handler.go

```diff
@@ -49,29 +63,49 @@ type ObservationStore interface {
 
 // FileStore provides storage for filtered output files.
 type FileStore interface {
-	PutFile(reader io.Reader, filter *observation.Filter) (url string, err error)
+	PutFile(reader io.Reader, fileID string) (url string, err error)
 }
 
 // Producer handles producing output events.
 type Producer interface {
-	CSVExported(filterID, fileURL string) error
+	CSVExported(e *CSVExported) error
+}
+
+type FileIDGenerator interface {
+	NewID() string
 }
 
 // Handle the export of a single filter output.
 func (handler *ExportHandler) Handle(event *FilterSubmitted) error {
+	var csvExported *CSVExported
+	var err error
+
+	if event.FilterID != "" {
+		csvExported, err = handler.filterJob(event)
+	} else {
+		csvExported, err = handler.prePublishJob(event)
+	}
 
-	filter, err := handler.filterStore.GetFilter(event.FilterID)
 	if err != nil {
 		return err
 	}
 
-	log.Debug("filter retrieved for output", log.Data{
-		"instance_id": filter.InstanceID,
-		"filter_id":   filter.FilterID})
+	if err := handler.eventProducer.CSVExported(csvExported); err != nil {
+		return errors.Wrap(err, "error while attempting to send csv exported event to producer")
+	}
+	return nil
+}
+
+func (handler *ExportHandler) filterJob(event *FilterSubmitted) (*CSVExported, error) {
+	log.Info("handling filter job", log.Data{"filter_id": event.FilterID})
+	filter, err := handler.filterStore.GetFilter(event.FilterID)
+	if err != nil {
+		return nil, err
+	}
 
 	csvRowReader, err := handler.observationStore.GetCSVRows(filter, nil)
```

Comment:
`csvRowReader ` does't seem to be closed?

Reply:
added